### PR TITLE
Add support to unstructured resources to predicates

### PR DESCRIPTION
**Pull request by davidebianchi**
## Description

Initial work to fix stale cache issues, as we'll need to convert all watches to possibly use unstructured resources.

For this initial work, add support to unstructured resources to predicates, and added unit tests.

From those unit tests, there are some predicates that will need a review, since it's not clear if it's correct to listen specific changes (those predicates are flagged with `FIXME` comment).

Jira task: initial work to https://issues.redhat.com/browse/RHOAIENG-47119

## How Has This Been Tested?
Unit testing

## Merge criteria

- [x] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [x] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [x] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work
- [x] The developer has run the integration test pipeline and verified that it passed successfully

### E2E test suite update requirement

When bringing new changes to the operator code, such changes are by default required to be accompanied by extending and/or updating the E2E test suite accordingly.

To opt-out of this requirement:
1. **Please inspect the [opt-out guidelines](https://github.com/opendatahub-io/opendatahub-operator/blob/main/docs/e2e-update-requirement-guidelines.md)**, to determine if the nature of the PR changes allows for skipping this requirement
2. If opt-out is applicable, provide justification in the dedicated `E2E update requirement opt-out justification` section below
3. Check the checkbox below:
- [x] Skip requirement to update E2E test suite for this PR
4. Submit/save these changes to the PR description. This will automatically trigger the check.

#### E2E update requirement opt-out justification
Updated predicates library, adding tests and support to unstructured resources. No feature changes.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved detection of changes for Deployments, ConfigMaps, Secrets, HTTPRoute references, and DataScienceCluster across typed and unstructured resources; handles missing or partial status/data more gracefully.
  * Removed legacy readiness predicate to avoid stale readiness behavior.

* **New Features**
  * Added exported predicates for ConfigMap and Secret content changes and enhanced DataScienceCluster and HTTPRoute handling across formats.

* **Tests**
  * Added extensive predicate tests covering structured/unstructured inputs, edge cases, name/namespace scenarios, and lifecycle events.

### CODE REVIEW

File: pkg/controller/predicates/resources/resources_test.go [hunk 1]
```diff
@@ -87,3 +98,1102 @@ func TestAnnotationChanged(t *testing.T) {
...
+	runNamePredicateTests(t, predicate, namePredicateTestConfig{
+		matchingName:    "test-name",
+		nonMatchingName: "other-name",
+		testCreate:      true,
+		testUpdate:      true,
+		testDelete:      false, // Delete always returns true
+		testGeneric:     false, // Generic always returns true
+	})
+
+	// FIXME: is correct to always return true with Delete predicate?
```
Comment (davidebianchi) [inline pkg/controller/predicates/resources/resources_test.go:886]: Is this predicate working as expected? From `CreatedOrUpdatedName` name, I'd expect this to be false.

File: pkg/controller/predicates/resources/resources_test.go [hunk 2]
```diff
@@ -87,3 +98,1102 @@ func TestAnnotationChanged(t *testing.T) {
...
+	t.Run("Delete", func(t *testing.T) {
+		t.Parallel()
+
+		g := NewWithT(t)
+
+		g.Expect(predicate.Delete(event.DeleteEvent{})).To(BeTrue())
+	})
+
+	// FIXME: is correct to always return true with Generic predicate?
```
Comment (davidebianchi) [inline pkg/controller/predicates/resources/resources_test.go:895]: Is this predicate working as expected? From `CreatedOrUpdatedName` name, I'd expect this to be false.

File: pkg/controller/predicates/resources/resources_test.go [hunk 3]
```diff
@@ -87,3 +98,1102 @@ func TestAnnotationChanged(t *testing.T) {
...
+	runNamePredicateTests(t, pred, namePredicateTestConfig{
+		matchingName:    "test-name",
+		nonMatchingName: "other-name",
+		testCreate:      true,
+		testUpdate:      true,
+		testDelete:      true,
+		testGeneric:     false, // Generic always returns true
+	})
+
+	// FIXME: is correct to always return true with Generic predicate?
```
Comment (davidebianchi) [inline pkg/controller/predicates/resources/resources_test.go:919]: Is this predicate working as expected? From `CreatedOrUpdatedOrDeletedNamed` name, I'd expect this to be false.

File: pkg/controller/predicates/resources/resources_test.go [hunk 4]
```diff
@@ -87,3 +98,1102 @@ func TestAnnotationChanged(t *testing.T) {
...
+	runNamePredicateTests(t, pred, namePredicateTestConfig{
+		matchingName:    "test-pod",
+		nonMatchingName: "other-pod",
+		testCreate:      true,
+		testUpdate:      true,
+		testDelete:      true,
+		testGeneric:     false, // Generic always returns true
+	})
+
+	// FIXME: is correct to always return true with Generic predicate?
```
Comment (davidebianchi) [inline pkg/controller/predicates/resources/resources_test.go:943]: Is this predicate working as expected? From `CreatedOrUpdatedOrDeletedNamePrefixed` name, I'd expect this to be false.

File: pkg/controller/predicates/resources/resources_test.go [hunk 5]
```diff
@@ -87,3 +98,1102 @@ func TestAnnotationChanged(t *testing.T) {
...
+		g.Expect(predicate.Delete(event.DeleteEvent{
+			Object: &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: "gateway-cert"}},
+		})).To(BeTrue())
+
+		g.Expect(predicate.Delete(event.DeleteEvent{
+			Object: &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: "other-secret"}},
+		})).To(BeFalse())
+	})
+
+	// FIXME: is correct to always return true with Generic predicate?
```
Comment (davidebianchi) [inline pkg/controller/predicates/resources/resources_test.go:1004]: Also this, is not defined. Is it correct to manage all generic event?

File: pkg/controller/predicates/resources/resources_test.go [hunk 6]
```diff
@@ -87,3 +98,1102 @@ func TestAnnotationChanged(t *testing.T) {
...
+			})).To(BeFalse())
+
+			g.Expect(predicate.Update(event.UpdateEvent{
+				ObjectOld: unstructuredPod(t, g, &corev1.Pod{ObjectMeta: metav1.ObjectMeta{ResourceVersion: "1", Generation: 1}}),
+				ObjectNew: unstructuredPod(t, g, &corev1.Pod{ObjectMeta: metav1.ObjectMeta{ResourceVersion: "1", Generation: 1}}),
+			})).To(BeFalse())
+		})
+	})
+
+	// FIXME: is correct to always return true with Generic predicate?
```
Comment (davidebianchi) [inline pkg/controller/predicates/resources/resources_test.go:1087]: Also this, is not defined. Is it correct to manage all generic event?

File: pkg/controller/predicates/resources/resources.go [hunk 7]
```diff
@@ -64,36 +117,110 @@ func Deleted() predicate.Funcs {
...
+		return !reflect.DeepEqual(oldData, newData)
 	},
 }
 
+// FIXME: is this function correct? By default CreateFunc and UpdateFunc are true.
 var DSCDeletionPredicate = predicate.Funcs{
 	DeleteFunc: func(e event.DeleteEvent) bool {
 		return true
 	},
 }
```
Comment (davidebianchi) [inline pkg/controller/predicates/resources/resources.go:176]: Is this predicate correct?

File: pkg/controller/predicates/resources/resources.go [hunk 8]
```diff
@@ -64,36 +117,110 @@ func Deleted() predicate.Funcs {
...
-		oldSecret, _ := e.ObjectOld.(*corev1.Secret)
-		newSecret, _ := e.ObjectNew.(*corev1.Secret)
-		return !reflect.DeepEqual(oldSecret.Data, newSecret.Data)
+		oldData, oldOk := getSecretData(e.ObjectOld)
+		newData, newOk := getSecretData(e.ObjectNew)
+		if !oldOk || !newOk {
+			return false
+		}
+		return !reflect.DeepEqual(oldData, newData)
 	},
```
[the remark this answers is not in the document]
Comment (davidebianchi) [reply pkg/controller/predicates/resources/resources.go]: Is it possible that old/new differ in representation? It should never happen

File: pkg/controller/predicates/resources/resources.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (davidebianchi) [reply pkg/controller/predicates/resources/resources.go:176]: Already add a comment to highlight this

File: pkg/controller/predicates/resources/resources.go [hunk 9]
```diff
@@ -1,46 +1,99 @@
...
 
 var _ predicate.Predicate = DeploymentPredicate{}
 
+// hasGVK checks if an unstructured object matches the expected GroupVersionKind.
+func hasGVK(u *unstructured.Unstructured, expected schema.GroupVersionKind) bool {
+	objGVK := u.GetObjectKind().GroupVersionKind()
+	return objGVK.Group == expected.Group &&
+		objGVK.Version == expected.Version &&
+		objGVK.Kind == expected.Kind
+}
```
[the remark this answers is not in the document]
Comment (davidebianchi) [reply pkg/controller/predicates/resources/resources.go:29]: This is now improved, since the expected type is unstructured, we could avoid the conversion

File: pkg/controller/predicates/resources/resources.go [hunk 10]
```diff
@@ -1,46 +1,93 @@
...
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/cluster/gvk"
 	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/resources"
 )
 
 var _ predicate.Predicate = DeploymentPredicate{}
 
+// hasGVK checks if an unstructured object matches the expected GroupVersionKind.
+func hasGVK(u *unstructured.Unstructured, expected schema.GroupVersionKind) bool {
+	objGVK := u.GetObjectKind().GroupVersionKind()
+	return objGVK.Group == expected.Group &&
```
Comment (lburgazzoli) [inline pkg/controller/predicates/resources/resources.go]: nit: you can probably compare the GVKs directly `objGVK == expected`

File: pkg/controller/predicates/resources/resources.go [hunk 11]
```diff
@@ -64,36 +111,102 @@ func Deleted() predicate.Funcs {
...
+	if ok {
+		return cm.Data, true
+	}
+
+	u, ok := obj.(*unstructured.Unstructured)
+	if !ok || !hasGVK(u, gvk.ConfigMap) {
+		return nil, false
+	}
+
+	data, _, err := unstructured.NestedStringMap(u.Object, "data")
```
Comment (lburgazzoli) [inline pkg/controller/predicates/resources/resources.go]: wonder if in this case it would be better to use `NestedFieldNoCopy` so we can avoid copies

File: pkg/controller/predicates/resources/resources.go (hunk 7 above)
Comment (lburgazzoli) [reply pkg/controller/predicates/resources/resources.go:176]: I don't think so, according to [controller-runtime' s code ](https://github.com/kubernetes-sigs/controller-runtime/blob/8ebd0ffa23d359590c9c4c9079062e362e709b74/pkg/predicate/predicate.go#L64-L108), TypedPredicate return true in every function. 

Also,. we should probably revisit the need for the DSCI to watch the DSC (probably legacy observability), seems incorrect.

File: pkg/controller/predicates/resources/resources.go (hunk 10 above)
Comment (davidebianchi) [reply pkg/controller/predicates/resources/resources.go]: Yes, updated to use this comparison!

File: pkg/controller/predicates/resources/resources.go (hunk 11 above)
Comment (davidebianchi) [reply pkg/controller/predicates/resources/resources.go]: It was difficult this improvement because should be manually converted to have the same return type of the typed resource.
I changed the implementation to improve the performance of this check, managing the types independently

Comment (lburgazzoli) [reply pkg/controller/predicates/resources/resources.go]: wonder if the method should just return any, does not see we actually need to be a map

File: pkg/controller/predicates/resources/resources.go (hunk 7 above)
Comment (davidebianchi) [reply pkg/controller/predicates/resources/resources.go:176]: Yes, this default behavior is present also in other predicates (I added some comments in the PR to highlight the behavior). The predicates are: 
- `CreatedOrUpdatedName`: Delete and Generic events returns always true
- `CreatedOrUpdatedOrDeletedNamed`: Generic event returns always true
- `CreatedOrUpdatedOrDeletedNamePrefixed`: Generic event returns always true
- `GatewayCertificateSecret`: Generic event returns always true
- `GatewayStatusChanged`: Generic event returns always true

Should those be updated to always return false?

For this one it invoke with some specific conditions the `configureManagedMonitoring` function. Should I update it to use the `Deleted` predicate?

File: pkg/controller/predicates/resources/resources.go (hunk 11 above)
Comment (davidebianchi) [reply pkg/controller/predicates/resources/resources.go]: Yes, you are right 😄 I prefer to keep the cm/secret functions separated, updated.

File: pkg/controller/predicates/resources/resources.go (hunk 7 above)
Comment (lburgazzoli) [reply pkg/controller/predicates/resources/resources.go:176]: I think it should probably revisit it later, to my understanding a generic event is generated from sources outside of kubernetes, so it is should not be an issue at this stage
